### DISCUSSION TITLE
How to handle circular imports in SQLAlchemy

### DISCUSSION THREAD

**Discussion by gabriel-frattini** [sqlalchemy/sqlalchemy#9576] [Usage Questions]
I'm having an issue with circular imports in SQLAlchemy. 

I have two files foo.py and bar.py. foo.py defines a SQLAlchemy class Foo, and bar.py defines a class Bar. 
Both Foo and Bar are each other's foreign keys, so they need to import each other. However, this is causing a circular import error. 

What's the best way to handle this issue? What are some general best practices for dealing with circular imports in SQLAlchemy?

```python
# foo.py

from sqlalchemy import Column, Integer, ForeignKey
from sqlalchemy.orm import relationship
from .bar import Bar

class Foo(Base):
    __tablename__ = 'foo'
    id = Column(Integer, primary_key=True)
    bar_id = Column(Integer, ForeignKey('bar.id'))
    bar: Mapped["Bar"] = relationship('Bar')

# bar.py

from sqlalchemy import Column, Integer, ForeignKey
from sqlalchemy.orm import relationship
from .foo import Foo

class Bar(Base):
    __tablename__ = 'bar'
    id = Column(Integer, primary_key=True)
    foo_id = Column(Integer, ForeignKey('foo.id'))
    foo: Mapped["Foo"] = relationship('Foo')
    ```

**Comment by zzzeek** [comment]
there is no circular import issue intrinsic with the above mapping, the classes refer to each other using a string name.  your program would want to import "foo" and "bar" into a third module and the code as written above should have no issue.

**Reply by gabriel-frattini** [reply]
The classes mapps each other with `Mapped["Foo"]`. So the classes refer to each other not only by string but the class itself also.

If I remove the foo import and keep `Mapped["Foo"]`. I get an error that Foo is not defined.

If I just remove `Mapped["Foo"]` I dont get a problem, but then I don't get the type safety I want, is that the trade off I have to make?

**Reply by zzzeek** [reply]
for python typing you use a TYPE_CHECKING block to indicate imports local to type checking:

https://mypy.readthedocs.io/en/stable/runtime_troubles.html#import-cycles

**Reply by mj2068** [reply]
Hi Mike, I have a similar question regarding:
> your program would want to import "foo" and "bar" into a third module

want to ask about best practices.
As I understand it, in order for the relationship to work, `Foo` and `Bar` must be imported before any actual operation happens. I think it's for the registry to aware the existence of mapped classes.
But, say in my code, I never used `Bar` explicitly, maybe I just want to `foo1.bar` or something.
Should I just have something like `from .bar import Bar` hanging somewhere in the code? what's best to do here?

**Reply by zzzeek** [reply]
if you have Foo / Bar and some code that uses them, the code that uses them would have imported everything needed to get Foo/Bar invoked at the class level.    Usually when the program starts up there's some main import that imports all model files.

**Comment by satk0** [comment]
TL;DR, based on @zzzeek comments, adding `if TYPE_CHECKING:` condition before each such an import solves the problem:

```python
# foo.py

from typing import TYPE_CHECKING

from sqlalchemy import Column, Integer, ForeignKey
from sqlalchemy.orm import relationship

if TYPE_CHECKING:
    from .bar import Bar

class Foo(Base):
    __tablename__ = 'foo'
    id = Column(Integer, primary_key=True)
    bar_id = Column(Integer, ForeignKey('bar.id'))
    bar: Mapped["Bar"] = relationship('Bar')


# bar.py

from typing import TYPE_CHECKING

from sqlalchemy import Column, Integer, ForeignKey
from sqlalchemy.orm import relationship

if TYPE_CHECKING:
    from .foo import Foo

class Bar(Base):
    __tablename__ = 'bar'
    id = Column(Integer, primary_key=True)
    foo_id = Column(Integer, ForeignKey('foo.id'))
    foo: Mapped["Foo"] = relationship('Foo')
```

**Reply by lupodellasleppa** [reply]
This works as long as you only use the class in type hints. If you need to reference the class itself (say, `order_by=Foo.id`) then importing the class in the `TYPE_CHECKING` block wouldn't suffice.

EDIT: A workaround for this is to use the `order_by` as string (e.g. `order_by="[Foo.id, Foo.name]"`), and in the case `Foo` is used in methods of `Bar`, import `Foo` inside the methods themselves

**Reply by DaanDL** [reply]
Honestly, I'm baffled that there is no clear cut solution to this in the documentation anywhere. It's such a basic scenario.

**Reply by CaselIT** [reply]
I'm not sure what you are looking for explicitly, but the string syntax is used multiple times in the docs, and the argument that support it document that

**Reply by DaanDL** [reply]
But if I only do the string syntax, without using TYPE_CHECKING magic, I get a warning in PyCharm.

**Reply by CaselIT** [reply]
Probably somewhere it could be mentioned in the docs more explicitly, but what you are describing is how python typing works, not something related to sqlalchemy. That's probably why it not explicitly stated in the docs

**Reply by zzzeek** [reply]
using TYPE_CHECKING blocks for top level imports used for typing only is not "magic", it is standard, ubiquitous best practice when using typing.  the documentation for this approach as mentioned above is in the mypy documentation as well as in [pep-484 itself](https://peps.python.org/pep-0484/#runtime-or-type-checking).  SQLAlchemy's typing features assume basic familiarity with Python typing.

**Reply by DaanDL** [reply]
Ok, got it!
